### DISCUSSION TITLE
[Dev Tools] Foundry compilation failure

### DISCUSSION THREAD

**Discussion by cytadela8** [zkSync-Community-Hub/zksync-developers#822] [Dev Tooling] [answered]
### Environment

Testnet

### Select the Dev Tool you are using

Foundry

### Provide a brief description of the functionality you're trying to implement and the issue you are running into.

Compilation of my contract (attached below). Fails with this error on zksync-foundry:

```
forge build --zksync
[⠃] Using zksolc-1.5.7
[⠑] Compiling (zksync)
[⠔] Compiling 91 files with zksolc and ZKsync solc 0.8.28
[⠒] zksolc and ZKsync solc 0.8.28 finished in 732.56ms
Error: Compiler run failed:

Error (2973)
TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 0.
 --> src/OptionResolver.sol:7:21:
  |
8 |     constructor ()  Ownable(msg.sender) {
  |                     ^^^^^^^^^^^^^^^^^^^

```

Contract code:
```
// SPDX-License-Identifier: None
pragma solidity ^0.8.28;

import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
import {Ownable2Step} from "@openzeppelin/contracts/access/Ownable2Step.sol";

contract OptionResolver is Ownable2Step {
    constructor ()  Ownable(msg.sender) {

    }

}
```

I'm using OZ with commit hash `dbb6104ce834628e473d2173bbc9d47f81a9eec3`

### Additional Details

Issue posted on behalf of client.

**Comment by alhenaai** [comment]
It appears that the error you're encountering during the compilation of your contract using Foundry on ZKsync is due to a misuse of the `Ownable` constructor. In the OpenZeppelin `Ownable` contract, the constructor does not take any arguments, but in your contract code, you are trying to pass `msg.sender` as an argument to the `Ownable` constructor.

Here's how you can correct the issue:

1. **Remove the Argument from the Constructor Call**: Since the `Ownable` constructor does not expect any arguments, you should not pass `msg.sender` to it. The `Ownable` contract automatically sets the contract deployer as the owner.

Modify your contract code as follows:

```solidity
// SPDX-License-Identifier: None
pragma solidity ^0.8.28;

import "@openzeppelin/contracts/access/Ownable.sol";
import "@openzeppelin/contracts/access/Ownable2Step.sol";

contract OptionResolver is Ownable2Step {
    constructor () Ownable() {
        // Constructor body if needed
    }
}
```

2. **Ensure Correct Inheritance**: If `OptionResolver` is intended to use features from both `Ownable` and `Ownable2Step`, ensure that these contracts are compatible and that you're using them as intended. Typically, `Ownable2Step` might already include the functionality of `Ownable`, so double-check if you need to inherit from both.

3. **Recompile the Contract**: After making the necessary changes, try recompiling the contract using the `forge build --zksync` command.

If you continue to face issues or if the error persists, it might be helpful to review the OpenZeppelin documentation or the specific versions of the contracts you are using to ensure compatibility with your Solidity version and constructor expectations.

For more detailed discussions and support, you can also visit the [ZKsync developers discussions on Github](https://github.com/zkSync-Community-Hub/zksync-developers/discussions).

Your feedback helps us serve you better. If this fully addresses your question, please give it a thumbs up. If it doesn't, please give it a thumbs down. This will inform the community and team so we can improve our support.

**Comment by dutterbutter** [comment]
Hello @cytadela8 

This is based on your OZ dependency version. Please double check the version being used. Here is an example of the latest OZ dep being used in a foundry project with `Ownable`.

- https://github.com/dutterbutter/debug-paymaster-input

**Comment by dutterbutter** [comment]
I will look to create a min-example for a quick test and post back here. Also is that version required?

**Comment by dutterbutter** [comment]
Here is a min-example that compiles fine with OZ version `dbb6104ce834628e473d2173bbc9d47f81a9eec3` 

- https://github.com/dutterbutter/min-example-ownable

Steps to run:

- `git clone git@github.com:dutterbutter/min-example-ownable.git` 
- `cd min-example-ownable` 
- `forge install` 
- `forge build --zksync` 

Foundry-zksync version:

```
forge --version
forge 0.0.2 (85f68fd 2024-12-10T00:35:43.371219000Z)
```

Perhaps can you add more details to the issue or try running `forge clean` and remove any cache and try compiling again?

**Comment by 21a1ss3** [comment]
We found one more required step. You need to run: `forge install velodrome-finance/contracts.git` in your example.
Now:
```
[⠃] Using zksolc-1.5.7
[⠊] Compiling (zksync)
[⠒] Compiling 32 files with zksolc and ZKsync solc 0.8.28
[⠑] zksolc and ZKsync solc 0.8.28 finished in 444.11ms
Error: Compiler run failed:
Error (2973)
TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 0.
 --> src/Ownable.sol:9:19:
  |
9 |     constructor() Ownable(msg.sender) {}
  |                   ^^^^^^^^^^^^^^^^^^^

Error (2973)
TypeError: Wrong argument count for modifier invocation: 1 arguments given but expected 0.
  --> src/Ownable2Step.sol:10:19:
   |
10 |     constructor() Ownable(msg.sender) {}
   |                   ^^^^^^^^^^^^^^^^^^^
```

**Reply by dutterbutter** [reply]
@21a1ss3 this is due to `velodrome-finance/contracts` making use of [v4](https://github.com/OpenZeppelin/openzeppelin-contracts/tree/7b3e7b7055d6e1226d4920ca7d7c446c0c0bfff5) of openzeppelin-contracts. Either update the dependency in `velodrome-finance/contracts` to v5 or make use of v4 entirely.

To be clear, two different major versions of openzeppelin-contracts, v4 and v5 are being used. This will result in a compilation error for both foundry and foundry-zksync and is not a good practice. I recommend upgrading deps to v5 of openzeppelin-contracts if possible. If not possible, then use v4 consistently.

**Reply by 21a1ss3** [reply]
This is a completely wrong advise.

First of all - I cannot do that because velodrome is a 3rd party code which I do not maintain myself and it has been compiled as is and got deployed by the maintainer team. Any changes there, changing the bytecode of already verified contracts. In my case I only using interfaces of it, so I do not directly using their libraries imports. And nevertheless standard Solidity&Forge handles this situation correctly.

Second thing, which is most probably most important implication what I can see now out of what we submitted earlier - why the wrong library even get used?  In my project, I imported OZ with version 5, which lies in the lib folder, but you using somehow library from library dependency for the main project. Which could be be catastrophic. For further explanation let me draw this layout of libs:

```
lib
  + Lib1 (version A)
  + Lib2 (version X)
     + Lib1 (version B)
```

From what I understood zksync compiler somehow using `Lib1 (version B)` for the main project. And this leads to following issues:
1) `Lib1` at version B either malfunctioning or has severe vulnerability, meanwhile keeping API compatible:
 
In this circumstances it is possible that `Lib2 `does not use broken portion of `Lib 1 (version B)`, but our project does, though in `Lib1 (version A)` everything is fixed & ok.

So given this facts we should get good & secure solution, but instead we unpredictably starting using vulnerable/broken code

2) Intended poisoning of dependencies:

Here I'm not sure how exactly libs names get resolved. Does your compiler use the repository name or only relies on folder name? In case if it is relies only on folder names, I can make a library, where I put poisoned version of dependency, which should never be used and even may be not linked within `Lib2` at all, but your compiler pick it up and start using within the main project. Which is very severe security issue after all.


So speaking overall, I would say this is very severe compiler bug, which, imo, should be addressed and fixed as regular solidity & forge behaves.

**Reply by dutterbutter** [reply]
I tested it against upstream `forge` / `solc` and you are right it does compile fine, my apologies. 

Let me look into this deeper for you and discuss with the compiler team.

**Reply by dutterbutter** [reply]
Okay so I think the issue is not with the compiler but rather how `foundry-zksync` is preparing the `standard_json_input`. I have flagged this and will begin working on resolving the issue. 

Thank you for posting 🙏  Will follow up here once fix is merged / released.

**Reply by dutterbutter** [reply, marked as answer]
Hey @21a1ss3 👋 

An update here, we've looked into the issue, and although we want to continue to investigate the compiler differences (evm / eravm) more closely I wanted to provide a solution in the meantime. 

I've [updated](https://github.com/dutterbutter/min-example-ownable/commit/5d990e64b684ed3cafd541b652d05194e13ade5e) the min-example: https://github.com/dutterbutter/min-example-ownable which compiles successfully with `velodrome-finance/contracts`. Specifically, updated the remapping so OZ v5 is used correctly. 

 Please check to see if this works for your use case and let me know. Thanks 🙏
